homeassistant/components/tuya/climate.py

```diff
@@ -0,0 +1,463 @@
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import TEMP_CELSIUS, TEMP_FAHRENHEIT
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .base import TuyaHaEntity
+from .const import (
+    DOMAIN,
+    TUYA_DEVICE_MANAGER,
+    TUYA_DISCOVERY_NEW,
+    TUYA_HA_DEVICES,
+    TUYA_HA_TUYA_MAP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+# Air Conditioner
+# https://developer.tuya.com/en/docs/iot/f?id=K9gf46qujdmwb
+DPCODE_SWITCH = "switch"
+DPCODE_TEMP_SET = "temp_set"
+DPCODE_TEMP_SET_F = "temp_set_f"
+DPCODE_MODE = "mode"
+DPCODE_HUMIDITY_SET = "humidity_set"
+DPCODE_FAN_SPEED_ENUM = "fan_speed_enum"
+
+# Temperature unit
+DPCODE_TEMP_UNIT_CONVERT = "temp_unit_convert"
+DPCODE_C_F = "c_f"
+
+# swing flap switch
+DPCODE_SWITCH_HORIZONTAL = "switch_horizontal"
+DPCODE_SWITCH_VERTICAL = "switch_vertical"
+
+# status
+DPCODE_TEMP_CURRENT = "temp_current"
+DPCODE_TEMP_CURRENT_F = "temp_current_f"
+DPCODE_HUMIDITY_CURRENT = "humidity_current"
+
+SWING_OFF = "swing_off"
+SWING_VERTICAL = "swing_vertical"
+SWING_HORIZONTAL = "swing_horizontal"
+SWING_BOTH = "swing_both"
+
+TUYA_HVAC_TO_HA = {
+    "hot": HVAC_MODE_HEAT,
+    "cold": HVAC_MODE_COOL,
+    "wet": HVAC_MODE_DRY,
+    "wind": HVAC_MODE_FAN_ONLY,
+    "auto": HVAC_MODE_AUTO,
+}
+
+TUYA_ACTION_TO_HA = {
+    "off": CURRENT_HVAC_OFF,
+    "heating": CURRENT_HVAC_HEAT,
+    "cooling": CURRENT_HVAC_COOL,
+    "wind": CURRENT_HVAC_FAN,
+    "auto": CURRENT_HVAC_IDLE,
+}
+
+TUYA_SUPPORT_TYPE = {
+    "kt",  # Air conditioner
+    "qn",  # Heater
+    "wk",  # Thermostat
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up tuya climate dynamically through tuya discovery."""
+    _LOGGER.debug("climate init")
+
+    hass.data[DOMAIN][entry.entry_id][TUYA_HA_TUYA_MAP][
+        DEVICE_DOMAIN
+    ] = TUYA_SUPPORT_TYPE
+
+    async def async_discover_device(dev_ids):
+        """Discover and add a discovered tuya climate."""
+        _LOGGER.debug("climate add-> %s", dev_ids)
+        if not dev_ids:
+            return
+        entities = await hass.async_add_executor_job(
+            _setup_entities, hass, entry, dev_ids
+        )
+        async_add_entities(entities)
+
+    entry.async_on_unload(
+        async_dispatcher_connect(
+            hass, TUYA_DISCOVERY_NEW.format(DEVICE_DOMAIN), async_discover_device
+        )
+    )
+
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    device_ids = []
+    for (device_id, device) in device_manager.device_map.items():
+        if device.category in TUYA_SUPPORT_TYPE:
+            device_ids.append(device_id)
+    await async_discover_device(device_ids)
+
+
+def _setup_entities(
```

Comment:
Does this do I/O?

Reply:
No

Reply:
Then we don't need to schedule it on the executor.

Reply:
fixed